### PR TITLE
Add `safe_write_unicode` function to tools

### PR DESCRIPTION
**Pull request by MichaelYochpaz**
## Description
This PR adds a few changes to the `tools` module made originally in the [download-rewrite PR](https://github.com/demisto/demisto-sdk/pull/3295).
This PR adds the `safe_write_unicode` function, which is needed for the new `download` rewrite, since we open files in memory and can't use the previous `_read_file` function (which opens files from disk).
It also adds support for using `pathlib.Path` on the `get_yml_paths_in_dir` and `get_child_directories` functions, and a few other minor fixes & changes.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,7 @@
 # Changelog
 ## Unreleased
+* Replaced the `tools._read_file` function with a more generic `tools.safe_read_unicode` function.
+* Added `pathlib.Path` support to the `tools.get_yml_paths_in_dir` and `tools.get_child_directories` functions.
```
Comment (dorschw) [inline CHANGELOG.md:4]: please add the `Internal:` prefix

File: demisto_sdk/commands/common/tools.py [hunk 2]
```diff
@@ -1,11 +1,13 @@
+from __future__ import annotations
```
Comment (dorschw) [inline demisto_sdk/commands/common/tools.py:1]: why?

File: demisto_sdk/commands/common/tools.py [hunk 3]
```diff
@@ -758,36 +766,39 @@ def get_last_remote_release_version():
                 f'{exc_msg[exc_msg.find(">") + 3:-3]}.\n'
                 f"This may happen if you are not connected to the internet."
             )
-        logger.info(
-            f"[yellow]Could not get latest demisto-sdk version.\nEncountered error: {exc_msg}[/yellow]"
+        logger.warning(
+            f"Could not find the latest version of 'demisto-sdk'.\nError: {exc_msg}"
```
Comment (dorschw) [inline demisto_sdk/commands/common/tools.py:770]: why remove `yellow` as part of this PR?

**Review by dorschw — APPROVED** [review]

File: demisto_sdk/commands/common/tools.py (hunk 2 above)
Comment (MichaelYochpaz) [reply demisto_sdk/commands/common/tools.py:1]: It allows using `|` for parameters typing like [here](https://github.com/demisto/demisto-sdk/pull/3709/files#diff-19219bb189e9e2c49968f53fb5cad06444a41c9502d62d4dade43e864a87cf59R693) on Python versions that are 3.7+ (without it, it's only allowed on 3.10+).

File: demisto_sdk/commands/common/tools.py (hunk 3 above)
Comment (MichaelYochpaz) [reply demisto_sdk/commands/common/tools.py:770]: Using a `warning` log level already uses a yellow color automatically, I already checked and confirmed it.
These are just some remains from when we didn't do that automatically and then these tags were needed.

File: CHANGELOG.md (hunk 1 above)
Comment (MichaelYochpaz) [reply CHANGELOG.md:5]:
```suggestion
Internal: * Replaced the `tools._read_file` function with a more generic `tools.safe_read_unicode` function.
Internal: * Added `pathlib.Path` support to the `tools.get_yml_paths_in_dir` and `tools.get_child_directories` functions.
```
Like this?

**Pull request merged**